libraries/Parsers/Microsoft.Bot.Builder.Parsers.LU/Parser/NewEntitySection.cs

```diff
@@ -0,0 +1,232 @@
+            }
+
+            return null;
+        }
+
+        public List<SynonymElement> ExtractSynonyms(LUFileParser.NewEntitySectionContext parseTree)
+        {
+            var synonymsOrPhraseList = new List<SynonymElement>();
+            if (parseTree.newEntityDefinition().newEntityListbody() != null)
+            {
+                foreach (var errorItemStr in parseTree.newEntityDefinition().newEntityListbody().errorString())
+                {
+                    if (!string.IsNullOrEmpty(errorItemStr.GetText().Trim()))
+                    {
+                        Errors.Add(
+                            Diagnostic.BuildDiagnostic(
+                                message: "Invalid list entity line, did you miss '-' at line begin?",
+                                context: errorItemStr));
+                    }
+                }
+
+                var bodyElement = new SynonymElement();
+                foreach (var normalItemStr in parseTree.newEntityDefinition().newEntityListbody().normalItemString())
+                {
+                    var trimedItemStr = normalItemStr.GetText().Trim();
+                    var normalizedValueMatch = Regex.Match(trimedItemStr, @"(?: |\t)*-(?: |\t)*(.*)(?: |\t)*:$");
+                    if (normalizedValueMatch.Success)
+                    {
+                        if (bodyElement.NormalizedValue != null)
+                        {
+                            // This is not the first value in the list
+                            synonymsOrPhraseList.Add(bodyElement);
+                            bodyElement = new SynonymElement();
+                        }
+
+                        bodyElement.NormalizedValue = normalizedValueMatch.Groups[1].Value.Trim();
+                    }
+                    else
+                    {
+                        var splitedStr = trimedItemStr.Split('-');
+                        bodyElement.Synonyms.Add(splitedStr[1].Trim());
+                    }
+                }
+
+                if (bodyElement.NormalizedValue != null)
+                {
+                    // There was at least one
+                    synonymsOrPhraseList.Add(bodyElement);
+                }
+            }
+
+            if (!string.IsNullOrEmpty(Type) && Type.IndexOf('=') > -1 && synonymsOrPhraseList.Count == 0)
+            {
+                var errorMsg = $"no synonyms list found for list entity definition: \"{parseTree.newEntityDefinition().newEntityLine().GetText()}\"";
+                var error = Diagnostic.BuildDiagnostic(
+                    message: errorMsg,
+                    context: parseTree.newEntityDefinition().newEntityLine(),
+                    severity: DiagnosticSeverity.Warn);
+                Errors.Add(error);
+            }
+
+            return synonymsOrPhraseList;
+        }
+
+        public List<string> ExtractPhraseList(LUFileParser.NewEntitySectionContext parseTree)
```

Comment:
Should *all* of these List<T> be IList<T> ? I think so.

Reply:
The params on all the public methods should check for null and throw.

Reply:
I use the AddRange funtionality in a lot of places. I believe that IList doesnt provide that method, is there a work around.